runtime/query.go

```diff
@@ -45,362 +46,294 @@ func (*defaultQueryParser) Parse(msg proto.Message, values url.Values, filter *u
+			}
+			// Look for enum by number
+			v = enum.Descriptor().Values().ByNumber(protoreflect.EnumNumber(i))
+			if v == nil {
+				return protoreflect.Value{}, fmt.Errorf("could not find enum: %q", value)
+			}
+		}
+		return protoreflect.ValueOfEnum(v.Number()), nil
+	case protoreflect.Int32Kind, protoreflect.Sint32Kind, protoreflect.Sfixed32Kind:
+		v, err := strconv.ParseInt(value, 10, 32)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing integer: %w", err)
+		}
+		return protoreflect.ValueOfInt32(int32(v)), nil
+	case protoreflect.Int64Kind, protoreflect.Sint64Kind, protoreflect.Sfixed64Kind:
+		v, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing integer: %w", err)
+		}
+		return protoreflect.ValueOfInt64(v), nil
+	case protoreflect.Uint32Kind, protoreflect.Fixed32Kind:
+		v, err := strconv.ParseUint(value, 10, 32)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing unsigned integer: %w", err)
+		}
+		return protoreflect.ValueOfUint32(uint32(v)), nil
+	case protoreflect.Uint64Kind, protoreflect.Fixed64Kind:
+		v, err := strconv.ParseUint(value, 10, 64)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing unsigned integer: %w", err)
+		}
+		return protoreflect.ValueOfUint64(v), nil
+	case protoreflect.FloatKind:
+		v, err := strconv.ParseFloat(value, 32)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing float: %w", err)
+		}
+		return protoreflect.ValueOfFloat32(float32(v)), nil
+	case protoreflect.DoubleKind:
+		v, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing float: %w", err)
+		}
+		return protoreflect.ValueOfFloat64(v), nil
+	case protoreflect.StringKind:
+		return protoreflect.ValueOfString(value), nil
+	case protoreflect.BytesKind:
+		v, err := base64.StdEncoding.DecodeString(value)
+		if err != nil {
+			return protoreflect.Value{}, fmt.Errorf("parsing base64 bytes: %w", err)
 		}
-		f.Index(i).Set(result[0].Convert(f.Index(i).Type()))
+		return protoreflect.ValueOfBytes(v), nil
+	case protoreflect.MessageKind, protoreflect.GroupKind:
+		return parseMessage(msgValue.Get(fieldDesc).Message(), value)
+	default:
+		panic(fmt.Sprintf("unknown field kind: %v", fieldDesc.Kind()))
 	}
-	return nil
 }
 
-func populateField(f reflect.Value, value string, props *proto.Properties) error {
-	i := f.Addr().Interface()
+func parseMessage(msgValue protoreflect.Message, value string) (protoreflect.Value, error) {
+	msgDesc := msgValue.Descriptor()
 
 	// Handle protobuf well known types
-	var name string
-	switch m := i.(type) {
-	case interface{ XXX_WellKnownType() string }:
-		name = m.XXX_WellKnownType()
-	case proto.Message:
-		const wktPrefix = "google.protobuf."
-		if fullName := proto.MessageName(m); strings.HasPrefix(fullName, wktPrefix) {
-			name = fullName[len(wktPrefix):]
-		}
+	if msgDesc.FullName().Parent() != "google.protobuf" {
```

Comment:
Is `Parent` documented to not use "actual" fully-qualified names, ie `.google.protobuf`? You might have to handle this case as well.

Also technically, this doesn't check for WKT - you need to check the actual files. There's lists at various places, such as https://github.com/bufbuild/buf/blob/master/internal/gen/embed/wkt/wkt.gen.go, you could make a `map[string]struct{}` with the WKT file names.

Reply:
The protoreflect.FullName type is documented to never include the preceding "." by virtue of the type itself being called "FullName".

Reply:
> Also technically, this doesn't check for WKT - you need to check the actual files.

At least in the protobuf module itself, we only identify WKT by name. https://github.com/protocolbuffers/protobuf-go/blob/master/internal/detectknown/detect.go

Is there a case you're concerned about if we don't also check the names?

Reply:
I could reword the error message, because it doesn't cover all WKT (obviously, descriptors aren't here), and it doesn't try to. It handles the few we support right now. I'll remove this and make the type lookup more specific and fall through to the error in the switch instead.